File: tools/llm-wiki-core/wiki-search.ts
import fs from "node:fs/promises";
import path from "node:path";
import type { WikiSearchHit } from "./wiki-types.js";
import { getWikiDir } from "./paths.js";

const TOKEN_RE = /\w|[\uAC00-\uD7A3]/gu;
const SKIP = new Set(["queries", "maintenance"]);

function tokenize(text: string): string[] {
  const m = text.match(TOKEN_RE);
  if (!m) return [];
  return m.filter((t) => t.length > 1).map((t) => t.toLowerCase());
}

function splitSections(content: string): { heading: string; body: string }[] {
  const parts = content.split(/^##\s+/m);
  if (parts.length <= 1) return [{ heading: "", body: content }];
  const out: { heading: string; body: string }[] = [];
  const preamble = parts[0]?.trim();
  if (preamble) out.push({ heading: "", body: preamble });
  for (let i = 1; i < parts.length; i++) {
    const block = parts[i] ?? "";
    const nl = block.indexOf("\n");
    const heading = nl >= 0 ? block.slice(0, nl).trim() : block.trim();
    const body = nl >= 0 ? block.slice(nl + 1).trim() : "";
    out.push({ heading, body });
  }
  return out;
}

function shouldIndex(relParts: string[]): boolean {
  if (relParts.some((p) => SKIP.has(p))) return false;
  return relParts.some((p) => ["sources", "notes", "concepts"].includes(p));
}

function scoreChunk(
  queryTokens: string[],
  heading: string,
  body: string,
  relPath: string
): number {
  if (!queryTokens.length) return 0;
  const hay = `${relPath} ${heading} ${body}`.toLowerCase();
  const haySet = new Set(tokenize(hay));
  const qset = new Set(queryTokens);
  let overlap = 0;
  for (const t of qset) if (haySet.has(t)) overlap++;
  if (!overlap) return 0;
  let score = overlap / qset.size;
  const rl = relPath.toLowerCase();
  for (const t of queryTokens) {
    if (rl.includes(t)) score += 0.35;
    if (heading.toLowerCase().includes(t)) score += 0.25;
  }
  score += Math.min(body.length / 8000, 0.15);
  return score;
}

export async function searchWiki(
  query: string,
  topK = 8,
  minScore = 0.12
): Promise<WikiSearchHit[]> {
  const queryTokens = tokenize(query);
  if (!queryTokens.length) return [];

  const wikiDir = getWikiDir();
  const hits: WikiSearchHit[] = [];

  async function walk(dir: string) {
    let entries;
    try {
      entries = await fs.readdir(dir, { withFileTypes: true });
    } catch {
      return;
    }
    for (const ent of entries) {
      const full = path.join(dir, ent.name);
      if (ent.isDirectory()) {
        await walk(full);
        continue;
      }
      if (!ent.name.endsWith(".md") || ent.name.startsWith("_")) continue;
      const rel = path.relative(wikiDir, full).replace(/\\/g, "/");
      const relParts = rel.split("/");
      if (!shouldIndex(relParts)) continue;
      const content = await fs.readFile(full, "utf-8");
      for (const { heading, body } of splitSections(content)) {
        const sc = scoreChunk(queryTokens, heading, body, rel);
        if (sc < minScore) continue;
        const excerpt = body.slice(0, 1200) + (body.length > 1200 ? "…" : "");
        hits.push({
          relPath: rel,
          heading: heading || "(서문)",
          text: excerpt,
          score: sc,
        });
      }
    }
  }

  await walk(wikiDir);
  hits.sort((a, b) => b.score - a.score || a.relPath.localeCompare(b.relPath));

  const seen = new Set<string>();
  const unique: WikiSearchHit[] = [];
  for (const h of hits) {
    const key = `${h.relPath}:${h.heading}`;
    if (seen.has(key)) continue;
    seen.add(key);
    unique.push(h);
    if (unique.length >= topK) break;
  }
  return unique;
}

File: scripts/wiki_search.py
"""Local keyword search over llm-wiki markdown (stdlib only)."""

from __future__ import annotations

import re
from dataclasses import dataclass
from pathlib import Path


@dataclass
class WikiChunk:
    path: Path
    rel_path: str
    heading: str
    text: str
    score: float


_TOKEN_RE = re.compile(r"[\w가-힣]+", re.UNICODE)
_SKIP_PARTS = {"queries", "maintenance", ".git"}

SCOPE_SOURCES = "sources"
SCOPE_NOTES = "notes"
SCOPE_KNOWLEDGE = "knowledge"
SCOPE_ALL = "all"


def tokenize(text: str) -> list[str]:
    return [t.lower() for t in _TOKEN_RE.findall(text) if len(t) > 1]


def slugify(text: str, max_len: int = 48) -> str:
    s = text.lower()
    s = re.sub(r"[^a-z0-9가-힣]+", "-", s, flags=re.UNICODE)
    s = re.sub(r"-+", "-", s).strip("-")
    return (s[:max_len] or "query").rstrip("-")


def _should_index(path: Path, wiki_root: Path, scope: str) -> bool:
    if path.name.startswith("_"):
        return False
    if path.suffix.lower() != ".md":
        return False
    parts = set(path.relative_to(wiki_root).parts)
    if parts & _SKIP_PARTS:
        return False
    if scope == SCOPE_SOURCES:
        return "sources" in parts
    if scope == SCOPE_NOTES:
        return "notes" in parts
    if scope == SCOPE_KNOWLEDGE:
        return bool(parts & {"sources", "notes"})
    return True


def iter_wiki_files(wiki_root: Path, scope: str = SCOPE_KNOWLEDGE) -> list[Path]:
    files: list[Path] = []
    for p in wiki_root.rglob("*.md"):
        if _should_index(p, wiki_root, scope):
            files.append(p)
    return sorted(files)


def split_sections(content: str) -> list[tuple[str, str]]:
    parts = re.split(r"(?m)^##\s+", content)
    if not parts:
        return [("", content)]
    sections: list[tuple[str, str]] = []
    preamble = parts[0].strip()
    if preamble:
        sections.append(("", preamble))
    for block in parts[1:]:
        lines = block.split("\n", 1)
        title = lines[0].strip()
        body = lines[1].strip() if len(lines) > 1 else ""
        sections.append((title, body))
    return sections or [("", content)]


def score_chunk(query_tokens: list[str], heading: str, body: str, rel_path: str) -> float:
    if not query_tokens:
        return 0.0
    hay = f"{rel_path} {heading} {body}".lower()
    hay_tokens = set(tokenize(hay))
    qset = set(query_tokens)
    overlap = qset & hay_tokens
    if not overlap:
        return 0.0
    score = len(overlap) / len(qset)
    path_lower = rel_path.lower()
    for t in query_tokens:
        if t in path_lower:
            score += 0.35
        if t in heading.lower():
            score += 0.25
    score += min(len(body) / 8000, 0.15)
    return score


def search_wiki(
    wiki_root: Path,
    question: str,
    *,
    top_k: int = 8,
    min_score: float = 0.12,
    scope: str = SCOPE_KNOWLEDGE,
) -> list[WikiChunk]:
    query_tokens = tokenize(question)
    if not query_tokens:
        return []

    hits: list[WikiChunk] = []
    for path in iter_wiki_files(wiki_root, scope=scope):
        try:
            content = path.read_text(encoding="utf-8")
        except OSError:
            continue
        rel = path.relative_to(wiki_root).as_posix()
        for heading, body in split_sections(content):
            sc = score_chunk(query_tokens, heading, body, rel)
            if sc < min_score:
                continue
            excerpt = body[:1200] + ("…" if len(body) > 1200 else "")
            hits.append(
                WikiChunk(
                    path=path,
                    rel_path=rel,
                    heading=heading or "(서문)",
                    text=excerpt,
                    score=sc,
                )
            )

    hits.sort(key=lambda c: (-c.score, c.rel_path))
    seen: set[tuple[str, str]] = set()
    unique: list[WikiChunk] = []
    for h in hits:
        key = (h.rel_path, h.heading)
        if key in seen:
            continue
        seen.add(key)
        unique.append(h)
        if len(unique) >= top_k:
            break
    return unique
